api/wasm/cpp/proxy_wasm_intrinsics.h

```diff
@@ -1017,3 +1206,194 @@ inline Histogram<Tags...>* Histogram<Tags...>::New(std::string_view name,
+  std::string serialized_request;
+  request.SerializeToString(&serialized_request);
+  return proxy_grpcCall(service.data(), service.size(), service_name.data(), service_name.size(), method_name.data(), method_name.size(),
+      serialized_request.data(), serialized_request.size(), timeout_milliseconds);
+                         
+}
+
+inline uint32_t grpcStream(std::string_view service, std::string_view service_name, std::string_view method_name) {
+  return proxy_grpcStream(service.data(), service.size(), service_name.data(), service_name.size(), method_name.data(), method_name.size());
+}
+
+inline void grpcCancel(uint32_t token) {
+  return proxy_grpcCancel(token);
+}
+
+inline void grpcClose(uint32_t token) {
+  return proxy_grpcClose(token);
+}
+
+inline void grpcSend(uint32_t token, std::string_view message, bool end_stream) {
+  return proxy_grpcSend(token, message.data(), message.size(), end_stream ? 1 : 0);
+}
+
+inline void Context::httpCall(std::string_view uri, const HeaderStringPairs& request_headers,
+    std::string_view request_body, const HeaderStringPairs& request_trailers,
+    uint32_t timeout_milliseconds, HttpCallCallback callback) {
+  auto token = ::httpCall(uri, request_headers, request_body, request_trailers, timeout_milliseconds);
+  if (token) {
+    http_calls_[token] = std::move(callback);
+  } else {
+    throw ProxyException("httpCall failed");
+  }
+}
+
+inline void Context::onHttpCallResponse(uint32_t token, std::unique_ptr<WasmData> header_pairs,
+    std::unique_ptr<WasmData> body, std::unique_ptr<WasmData> trailer_pairs) {
+  auto it = http_calls_.find(token);
+  if (it != http_calls_.end()) {
+    it->second(std::move(header_pairs), std::move(body), std::move(trailer_pairs));
+    http_calls_.erase(token);
+  }
+}
+
+inline void Context::grpcSimpleCall(std::string_view service, std::string_view service_name, std::string_view method_name,
+    const google::protobuf::MessageLite &request, uint32_t timeout_milliseconds, Context::GrpcSimpleCallCallback callback) {
+  auto token = grpcCall(service, service_name, method_name, request, timeout_milliseconds);
+  if (token) {
+    simple_grpc_calls_[token] = std::move(callback);
+  } else {
+    throw ProxyException("grpcCall failed");
+  }
+}
+
+inline void GrpcCallHandlerBase::cancel() {
+  grpcCancel(token_);
+  context_->grpc_calls_.erase(token_);
+}
+
+inline void GrpcStreamHandlerBase::reset() {
+  grpcCancel(token_);
+  context_->grpc_streams_.erase(token_);
+}
+
+inline void GrpcStreamHandlerBase::close() {
+  grpcClose(token_);
+  // NB: callbacks can still occur: reset() to prevent further callbacks.
+}
+
+inline void GrpcStreamHandlerBase::send(std::string_view message, bool end_of_stream) {
+  grpcSend(token_, message, end_of_stream);
+  if (end_of_stream) {
+    // NB: callbacks can still occur: reset() to prevent further callbacks.
+  }
+}
+
+inline void Context::onGrpcCreateInitialMetadata(uint32_t token) {
+  {
```

Comment:
You could avoid unnecessary scans here by branching on `IsGrpcCallToken(token)`.

Reply:
done